src/Pathy/Path.purs

```diff
@@ -0,0 +1,291 @@
+type AbsDir = Path Abs Dir
+
+-- | A type describing a file whose location is given relative to some other,
+-- | unspecified directory (referred to as the "current directory").
+type RelFile = Path Rel File
+
+-- | A type describing a file whose location is absolutely specified.
+type AbsFile = Path Abs File
+
+-- | The root directory, which can be used to define absolutely-located resources.
+rootDir :: Path Abs Dir
+rootDir = Init
+
+-- | The "current directory", which can be used to define relatively-located
+-- | resources.
+currentDir :: Path Rel Dir
+currentDir = Init
+
+-- | Creates a path which points to a relative file of the specified name.
+file :: NonEmptyString -> Path Rel File
+file = file' <<< Name
+
+-- | Creates a path which points to a relative file of the specified name.
+file' :: Name File -> Path Rel File
+file' = In currentDir
+
+-- | Creates a path which points to a relative directory of the specified name.
+dir :: NonEmptyString -> Path Rel Dir
+dir = dir' <<< Name
+
+-- | Creates a path which points to a relative directory of the specified name.
+dir' :: Name Dir -> Path Rel Dir
+dir' = In currentDir
+
+-- | Creates a path that points to the parent directory of the specified path.
+parentOf :: forall a. Path a Dir -> Path a Dir
+parentOf p = ParentOf p
+
+-- | Extends a path with a file or directory under the current path.
+extendPath :: forall a b. Path a Dir -> Name b -> Path a b
+extendPath p = In p
+
+-- | Given a directory path, appends a relative path to extend the original
+-- | path.
+appendPath :: forall a b. Path a Dir -> Path Rel b -> Path a b
+appendPath = case _, _ of
+  Init, Init -> Init
+  ParentOf p, Init -> ParentOf (p </> Init)
+  In p (Name d), Init -> In (p </> Init) (Name d)
+  p1, ParentOf p2 -> ParentOf (p1 </> p2)
+  p1, In p2 n -> In (p1 </> p2) n
+
+infixl 6 appendPath as </>
+
+-- | Ascends into the parent of the specified directory, then descends into
+-- | the specified path.
+-- |
+-- | ```purescript
+-- | canonicalize (rootDir </> dir "foo" <..> dir "bar") = rootDir </> dir "bar"
+-- | ```
+parentAppend :: forall a b. Path a Dir -> Path Rel b -> Path a b
+parentAppend d p = parentOf d </> p
+
+infixl 6 parentAppend as <..>
+
+-- | Canonicalizes a path, by reducing things in the form `/x/../` to just
+-- | `/x/`. Paths like `/../` will be normalized to `/`.
+canonicalize :: forall a b. IsRelOrAbs a => Path a b -> Path a b
+canonicalize p = fromMaybe p (go p)
+  where
+    go :: forall b'. Path a b' -> Maybe (Path a b')
+    go = case _ of
+      Init ->
+        Nothing
+      p'@(ParentOf Init) ->
+        foldRelOrAbs (const Nothing) (const (Just Init)) p'
+      ParentOf (In p' _) ->
+        -- Coercion is safe as `ParentOf` can only appear where `b' ~ Dir`
+        Just $ (unsafeCoerce :: Path a Dir -> Path a b') (canonicalize p')
+      ParentOf p' ->
+        case go p' of
+          Just p'' -> Just $ canonicalize (ParentOf p'')
+          Nothing -> Nothing
```

Comment:
We can also replace this entire `case go p' of...` with `canonicalize <<< ParentOf <$> go p'`.

Reply:
✅